[WIP] vscode-extensions: support multiple extension registries

ShamrockLee opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
According to the [Microsoft Visual Studio Marketplace Term of Use](https://cdn.vsassets.io/v/M146_20190123.39/_content/Microsoft-Visual-Studio-Marketplace-Terms-of-Use.pdf), VSCodium and editors other than the officially-built Visual Studio Code are prohibited from using "Marketplace Offerings" including extensions. These editors thus use the [Open VSX Registry](https://open-vsx.org/about) to provide extension searching/installation features instead of the [Visual Studio Marketplace](https://marketplace.visualstudio.com/vscode) .

~If applied, extensions will be downloaded from the Open VSX Registry by default to make them legally available to users of open-source vscode builds and other derived editors.~

This PR provides a framework to supply extensions from different registries. Four  registies, `vscode-marketplace`, `openvsx`, `upstream` and `standalone`, will be added.

Two nix-prefetch packages, `nix-prefetch-openvsx` and `nix-prefetch-vscode-mktplc` are included to make pre-fetching extensions. The output will be in JSON format similar to `nix-prefetch-github`.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on pkgs/misc/vscode-extensions/update_installed_exts.sh:
```diff
@@ -16,39 +16,98 @@ function clean_up() {
-    # I don't like 'rm -Rf' lurking in my scripts but this seems appropriate.
+    USE_MS_MKTPLC="false";
+    local FETCHED_JSON_STRING="";
+    if test $# -ge 3 && test -n $3 && test $3 = "true"; then
```
I would be happier if we use [[ instead of the old style test here.

ShamrockLee replied:
Would it be okay to use `[`?

ShamrockLee replied:
Now I see the necessity to use BASH (to propagate argument list correctly) and the advantage to use double bracket.
Suggestion applied.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/update_installed_exts.sh:
```diff
@@ -63,12 +122,22 @@ trap clean_up SIGINT
-    OWNER=$(echo "$i" | cut -d. -f1)
-    EXT=$(echo "$i" | cut -d. -f2)
+(
+    set +e
```
is there something in here which is likely to fail? Maybe we should just append a ``|| true`` to that command instead.

ShamrockLee replied:
The script is meant to run even if some extensions are not updated successfully.
Suggestions applied.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-vscode-mktplc/nix-prefetch-vscode-mktplc:
```diff
@@ -0,0 +1,94 @@
+SHA="$(nix-hash --flat --base32 --type sha256 "$EXTTMP/$N.zip")"
+
+# Clean up.
+rm -Rf "$EXTTMP"
```
```suggestion
rm -rf "$EXTTMP"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-vscode-mktplc/nix-prefetch-vscode-mktplc:
```diff
@@ -0,0 +1,94 @@
+URL="https://$PUBLISHER.gallery.vsassets.io/_apis/public/gallery/publisher/$PUBLISHER/extension/$NAME/$VER/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage"
+
+# Quietly but delicately curl down the file, blowing up at the first sign of trouble.
+curl --silent --show-error --fail -X GET -o "$EXTTMP/$N.zip" "$URL"
```
```suggestion
curl --silent --show-error --fail -o "$EXTTMP/$N.zip" "$URL"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-vscode-mktplc/nix-prefetch-vscode-mktplc:
```diff
@@ -0,0 +1,94 @@
+function clean_up() {
+    TDIR="${TMPDIR:-/tmp}"
+    echo "Script killed, cleaning up tempdirs: $TDIR/vscode_exts_*"
+    rm -Rf "$TDIR/vscode_exts_*"
```
```suggestion
    rm -rf "$TDIR/vscode_exts_*"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-vscode-mktplc/nix-prefetch-vscode-mktplc:
```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
```
```suggestion
#!/usr/bin/env bash
```

/bin/sh is bash on NixOS and something else on most other distros.

ShamrockLee replied:
Applied.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-vscode-mktplc/default.nix:
```diff
@@ -0,0 +1,44 @@
+    chmod +x $out/bin/${pname}
+  '';
+
+  postFixup = let
+    runtimePath = lib.makeBinPath buildInputs;
+  in ''
+    sed -i "2 i export PATH=${runtimePath}" "$out/bin/${pname}"
+  '';
```
```suggestion
  postFixup = ''
    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}" "$out/bin/${pname}"
  '';
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-openvsx/nix-prefetch-openvsx:
```diff
@@ -0,0 +1,97 @@
+SHA=$(nix-hash --flat --base32 --type sha256 "$EXTTMP/$N.zip")
+
+
+rm -Rf "$EXTTMP"
```
```suggestion
rm -rf "$EXTTMP"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-openvsx/nix-prefetch-openvsx:
```diff
@@ -0,0 +1,97 @@
+fi
+
+URL_VSIX="https://open-vsx.org/api/$PUBLISHER/$NAME/$VER/file/$N-$VER.vsix"
+curl --silent --show-error -X GET -L -o "$EXTTMP/$N.zip" "$URL_VSIX"
```
```suggestion
curl --silent --show-error -L -o "$EXTTMP/$N.zip" "$URL_VSIX"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-openvsx/nix-prefetch-openvsx:
```diff
@@ -0,0 +1,97 @@
+EXTTMP=$(mktemp -d -t vscode_exts_XXXXXXXX)
+URL_META="https://open-vsx.org/api/$PUBLISHER/$NAME/$VER"
+N_META="${N}_meta.json"
+curl --silent --show-error -X GET -o "$EXTTMP/$N_META" "$URL_META"
```
```suggestion
curl --silent --show-error -o "$EXTTMP/$N_META" "$URL_META"
```

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-openvsx/nix-prefetch-openvsx:
```diff
@@ -0,0 +1,97 @@
+            exit 0
+            ;;
+        *)
+            if test "$#" -lt 2; then
```
Please use [[ instead of the old test.

ShamrockLee replied:
Applied.

SuperSandro2000 commented on pkgs/misc/vscode-extensions/nix-prefetch-openvsx/default.nix:
```diff
@@ -0,0 +1,44 @@
+    chmod +x $out/bin/${pname}
+  '';
+
+  postFixup = let
+    runtimePath = lib.makeBinPath buildInputs;
+  in ''
+    sed -i "2 i export PATH=${runtimePath}" "$out/bin/${pname}"
+  '';
```
```suggestion
  postFixup = ''
    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}" "$out/bin/${pname}"
  '';
```

ShamrockLee commented:
Modified `nix-prefetch-openvsx` and `nix-prefetch-vscode-mktplc` to correctly delete the temporary directory.

The original way to handle exit in the `update_installed_exts.sh` only handles SIGINT (Ctrl-c) and successful exit, and will also delete the temporary directory used by another instance of this program.

ShamrockLee commented:
@SuperSandro2000 
It would be preferable to merge #113280 first and I'll then integrate with it.

1. `update_installed_exts.sh`
    *   #113280 changes the way argument lists are processed.
    *   This pr changes the way extensions are fetched. 
    *   The script needs integration.
2.  Documentation
    *   #113280 adds the documentation about VSCode and extensions.
    *   The changes in this PR needs to be documented, and the sha256 values of the examples need to be changed.
    *   Extra stuff to document will include:
        *   Sources of extensions (Open VSX Registry / official Marketplace)
        *   Two new prefetch scripts for extensions (`nix-prefetch-openvsx` and `nix-prefetch-vscode-mktplc`)

3.  Backward compatibility
    *   #113280 is totally backward compatible.
    *   This PR changes the default source of the extensions, and may affect the sha256 values of the extensions locally fetched by users.

SuperSandro2000 commented:
> It would be preferable to merge #113280 first and I'll then integrate with it.

Thats probably a good idea but I don't know the vscode package system that well to review the PRs well.

ShamrockLee commented:
Exposed `useMSMktplc` for future module making, with similar functionality to `meta.license.free`.

superherointj commented:
1) As hash differs between extension registries. And some people might choose to use MS VSCode (it's possible, right?) or even a new different registry might come along (since open-vsx is not neutral, [forbids GPL](https://www.eclipse.org/legal/licenses.php#nonapproved) chance exist etc.). Shouldn't extension hash be kept for each extension registry?

---

2) On my last attempted of reviewing, I stopped because I was having an issue on my host, when listing extensions using codium it would error: `codium --list-extensions`
```
Error: Trying to add a disposable to a DisposableStore that has already been disposed of. The added object will be leaked!
    at u.add (/nix/store/88jzhxkcdry21snb7x22zlk8xxl0kj46-vscodium-1.56.2/lib/vscode/resources/app/out/vs/code/node/cli.js:6:4076)
    at f._register (/nix/store/88jzhxkcdry21snb7x22zlk8xxl0kj46-vscodium-1.56.2/lib/vscode/resources/app/out/vs/code/node/cli.js:6:4479)
    at f.startWatching (/nix/store/88jzhxkcdry21snb7x22zlk8xxl0kj46-vscodium-1.56.2/lib/vscode/resources/app/out/vs/code/node/cliProcessMain.js:11:5879)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```
Seen this? Any suggestion?

ShamrockLee commented:
@superherointj

> since open-vsx is not neutral, forbids GPL etc

I didn't realize the possibility for GPL licensed program be prevented from being built/distributed by an Eclipse licensed platform. (Then why is it possible for GPL licensed project appears on the official marketplace or here on GitHub?)

> Shouldn't extension hash be kept for each extension registry?

The change of default source of extensions and the SHAs needs more discussions. It would still be an option to keep the default source and hashes, and document the way to override using the expressions generated by the prefetch (nix-prefetch-openvsx and nix-prefetch-vscode-mktplc) and update scripts.

It would also be possible to place the extensions from the OpenVSX separately, but it would probably increase the maintenance burden, and wouldn't help shorten the installation time, considering that most extensions are just downloaded and unpacked to the specified location.

> When listing extensions using codium it would error: codium --list-extensions

It doesn't happen on my NixOS laptop. Nevertheless, after integrating with #113280, the update script will no longer depend on `code-command --list-extensions`, and will read `provider.name` lines from the standard input instead.

superherointj commented:
Ideally we should mirror what actually exist in each registry. Meaning, it'd require an extension list per extension registry to accommodate for differences between registries (like names, ownership, versions, metadata, etc). This would also allow for new registries that might come along.

superherointj commented:
My expectations were:
* Have one nix file per registry containing all it's extensions.
* Importing extensions would be done as: `with registryAlias; [ ext_name ];`
* `useMSMktplc = true;` deemed unnecessary.
* new PR goal: `vscode-extensions: support multiple extension registry`

Makes sense? Thoughts? Am I overlooking something?

superherointj commented:
To reduce the review time this PR is taking, let's try to break it into parts.

Help me here to define the list of parts and order of it:

[WIP] Outline:
1) Define multiple registries abstraction.
2) Move current VSCode registry to the new abstraction. (No OpenVSX.)
3) CLI nix-prefetch-coderegistry-msmarketplace
4) CLI nix-prefetch-coderegistry-openvsx
5) Add OpenVSX.
6) Routines for automating updating of each registry.
7) ?
8) ?

Thoughts?

superherointj commented:
My thoughts/impressions on naming:
* 'extensions registry' seems a preferable name over 'marketplace' which could be a reference for a Microsoft product/marketing offering and it's not as a general name as the suggested and it also relates to trading which is not the case.
* I was considering 'code' to refer to both VSCode, Codium, or any other variant of it. But it did not seem right. Ideas?
* Ideally we should avoid unnecessary abbreviations. It hinders reading. A long name is preferable over a hardly readable abbreviation.
  * MKTPLC -> marketplace

Hope this is not too much nitpicking/bike shedding. :-)

ShamrockLee commented:
> Have one nix file per registry containing all it's extensions.

That's a great idea.

A straightforward way is to put the `registryRef` (current `mktplcRef`) or the hash value for one registry into a file, and put other build steps into another file.

The problems I have noticed when (manually) modifying the hashes of the current extension packages include:

*   The available versions in different registries may be different. Among cases with such cases, the newer versions are most likely provided by the official marketplace, and the latest version in the less-fresher registry would mostly (if not always) appear in the newer registry. Though not yet observed in the extensions, version changes may come with license/dependency/build method changes.
*   Very occasionally, the name (letter case) of the provider and extension in a very small amount of packages may be slightly different between registries. This can be resolved with attributes specific for nix, which is the approach (taken even before this PR) in `vscode-extensions/default.nix`. But for most of the packages, it sounds like unnecessary repetition.

ShamrockLee commented:
> Ideally we should avoid unnecessary abbreviations. It hinders reading. A long name is preferable over a hardly readable abbreviation.
> 
>     * MKTPLC -> marketplace

I agree.

I used mktplc simply because it is used everywhere here. I have thought that it were a convention to abbreviate like that.

superherointj commented:
> A straightforward way is to put the registryRef (current mktplcRef) or the hash value for one registry into a file, and put other build steps into another file.

Indeed, separating extension registry data and build steps/logic is ideal. I was trying to figure out if there was something else I could missing.

Relating extensions between registries cannot be guaranteed, is prone to errors, won't work. We should just faithfully replicate what actually exist on a registry. And leaving the user on charge of importing from the his desired registry.

SuperSandro2000 commented:
> 1. 1.2) Can [VSCode package](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vscode/vscode.nix?rgh-link-date=2021-05-27T17%3A21%3A54Z) still legally use Microsoft Marketplace?

We are just continue to use it because otherwise the vscode in NixOS is unusable when also managing extensions with nix.

Technically only the distributed version of VSCode from Microsoft is allowed to use it according to their Policy but there exist no technical measures that enforce that. If you are locally changing when base URLs from vsix to the Microsoft store everything will continue to work.

ShamrockLee committed:
nix-prefetch-vscode-marketplace, nix-prefetch-openvsx: init at 0.1.0

ShamrockLee commented:
1.  Thanks to the Open VSX Registry API, it is possible to download the metadata of all the extensions available on Open VSX Registry within a few seconds.
    There are currently 1131 packages inside the registry. Should they all be included?
2.  Should `allowSubstitute = false` be specified for extensions that doesn't require special build process? The main bottleneck when building most of the extensions are downloading, either from the registry or from the binary cache.
3.  Is there a VSCode Marketplace equivalence? I have heard of `extensionGallery.serviceUrl` and `extensionGallery.itemUrl`, but haven't found related documentation.

SuperSandro2000 commented:
> There are currently 1131 packages inside the registry. Should they all be included?

In nixpkgs? Probably not in file form because I am pretty sure most of them would be unused and the repo would grow by a lot. Maybe it could be done programmatically without listing them all?

> Should allowSubstitute = false be specified for extensions that doesn't require special build process? 

That wouldn't be a bad idea if the downloads are not yanked.

> Is there a VSCode Marketplace equivalence? I have heard of extensionGallery.serviceUrl and extensionGallery.itemUrl, but haven't found related documentation.

Not sure but those are the endpoints I know.

extensionGallery.itemUrl=https://marketplace.visualstudio.com/items
extensionGallery.serviceUrl=https://marketplace.visualstudio.com/_apis/public/gallery

ShamrockLee commented:
I'm working on the graduation requirements these two moths, so there would be little or no progress on this PR.

I plan to construct registries as attribute sets
```Nix
openvsx = {
  registry-reference-attrs = {
    # ...
    bbenoist.Nix = {
      publisher = "bbenoist";
      name = "Nix";
      version = "1.0.1";
      sha256 = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
    };
  };
  meta-attrs = {
    descriptions = "blablabla";
    homepage = "blablabla";
  };
}
```
and to add `openvsx.extensions` programmatically.

Some meta attributes can be extracted from `package.json` inside the VSIX file.

An issue would be long attribute chains:

In order to keep backward compatibility, extensions from vscode marketplace need to be injected under `vscode-extensions`. If the registries are also put under the `vscode-extension` attribute, there would be something like `pkgs.vscode-extensions.registries.vscode-marketplace.extensions`, which seems to be too long.

Would it be better to have `pkgs.vscode-registries` instead?

BTW, is it more preferable to use `registryRef` or `registry-reference`? `registryRefAttrs` or `registry-reference-attrs`?

superherointj reviewed (commented):
> I'm working on the graduation requirements these two moths, so there would be little or no progress on this PR.

No problem.

> In order to keep backward compatibility, extensions from vscode marketplace need to be injected under `vscode-extensions`. 

`vscode-extensions` should be an alias to Microsoft's registry to keep backward compatibility.
`vscode-registries` being the replacement. Other editors will make use of registries as well, eg. Onivim.
Once the new format is well set and communicated to users a decision on warning can be made - but no hurry.

> `pkgs.vscode-extensions.registries.vscode-marketplace.extensions`, which seems to be too long.

Don't worry about paths being too long. It's fine because VSCode is no different from other applications. Users have only to use `with` once. It's fine.

> BTW, is it more preferable to use `registryRef` or `registry-reference`? `registryRefAttrs` or `registry-reference-attrs`?

Why not `extensions`?

@SuperSandro2000 Any preferences on paths and naming here?

SuperSandro2000 commented:
> @SuperSandro2000 Any preferences on paths and naming here?

Not really right now. I am a vscode user but I don't use it on NixOS (yet).

superherointj commented:
@ShamrockLee How are your plans for this PR going?

ShamrockLee commented:
@superherointj sorry for the long delay.
I'll push up the current status of the registries (which is quite different from the current commits) this weekends.

superherointj commented:
@ShamrockLee Can you add comments to clarify purpose, inputs/outputs. (Like [here](https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix)). Specially at `registry-lib.nix`, `openvsx/default.nix`. Also run `nixpkgs-fmt` on files.

superherointj commented:
28.751 commits behind NixOS:master is quite a lot. Would be nice if you rebased to master.

superherointj commented:
Do you remember my comment on breaking this PR into phases?
Commits needs to be broken down, as in, commits should reflect the phases.
For example. The list of extensions of each registry should be a commit in itself.

ShamrockLee committed:
vscode-registries.openvsx: init

ShamrockLee committed:
vscode-registries.vscode-marketplace: init

ShamrockLee committed:
vscode-registry-commons: init

*   registry-lib: init
*   mkExtensionGeneral: init
*   modifiers: init

ShamrockLee committed:
vscode-registries: init

superherointj commented:
@ShamrockLee I suggest you add a new reviewer. I'm no longer active here. Sorry for that.

ShamrockLee commented:
@superherointj  Okay.

ShamrockLee commented:
@mohe2015 Thanks a lot!

> Also the full regeneration of the registry file seems to be extremely slow. Was that the case for you too?

IIRC, the `meta.platforms` are not set by default for most VSCode extension packages, so rebuild may happen only to some packages with custom build process. Beside, the main overhead to build most VSCode extensions comes from VSIX file downloading. For VSCode Marketplace users, the change should not take a long time to apply, since the SHA256 for existing extensions are kept unchanged.

mohe2015 commented:
Yeah I meant https://github.com/NixOS/nixpkgs/pull/121583/files#diff-ca7f7fcda6812c22c8b079f24d3379193a4120f03d137fe118c98d18a22348c5R17 but it seems like the download server is slow for me.

ShamrockLee commented:
It takes hours to download all the OpenVSX extensions (~1000 currently). I have filed a [feature request](https://github.com/eclipse/openvsx/issues/311) at eclipse/openvsx to enable SHA256 query, but haven't receive a response yet.

By the way, there are currently ~200 VSCode Marketplace packages under `vscode-extensions`, which will be moved to `vscode-registries.vscode-marketplace.extensions` and alias back to `vscode-extensions`.

ShamrockLee commented:
The above is all about initial works. There should be a more reasonable update approach.

mohe2015 commented:
`jq -r '.extensions[] | [ .namespace, .name, .version] | @sh | "nix-prefetch-openvsx " + . + "; echo ,"' ./all-extensions-raw.json | parallel -j 32 | tee -a registry-reference-list.json` seems to work better for me but the best thing would be upstream to provide the hash and we definitely need an incremental update script

mohe2015 commented on pkgs/misc/vscode-registries/openvsx/all-extensions-raw.json:
```diff
@@ -0,0 +1,19030 @@
+{
```
To my understanding this file can be removed? (Don't need to do that now)

ShamrockLee committed:
lib.meta.getLicenseFromSpdxId: drop warnings

Drop the warnings for lib.meta.getLicenseFromSpdxId fallback
to please the CI and for better user experience.

Create lib.meta.getLicenseFromSpdxId',
for which developers can specify the default value
and possible warning/error.

mohe2015 commented on pkgs/misc/vscode-registries/vscode-marketplace/default.nix:
```diff
@@ -0,0 +1,99 @@
+
+  base = final:
+    {
+      cacheUrl = "https://vscode.blob.core.windows.net/gallery/index";
```
Could it be that this url is heavily outdated or needs some parameters as https://vscode.blob.core.windows.net/gallery/index shows a heavily outdated version for e.g. https://marketplace.visualstudio.com/items?itemName=fterrag.vscode-php-cs-fixer ?

ShamrockLee replied:
What a bad news.

Now the revelant APIs for the VSCode Marketplace are only `serviceUrl` and `itemUrl`, whics are also what the VSCode-like editors use query the extensions. I haven't found any publicly available documentation about the query, but Open VSX has already implemented [a server-side relay]{https://github.com/eclipse/openvsx/wiki/Using-Open-VSX-in-VS-Code). Maybe they have already figured out the trick.

mohe2015 replied:
I think it is more of a legal issue and we are only allowed to support adding the extensions to the unmodified vscode to my limited understanding. The website https://marketplace.visualstudio.com/vscode contains "Download extension" links that at least show the link for individual extensions. The question is where you can get a list of extensions

ShamrockLee replied:
> I think it is more of a legal issue and we are only allowed to support adding the extensions to the unmodified vscode to my limited understanding.
That's why I open this PR. (The original purpose is to switch to Open VSX, and later superherontj suggests that it make more sense to support multiple registries)


The open-vsx people seems to know the trick, but neither they nor the Microsoft has published any documentation about the technical detail.

ShamrockLee commented:
Here's the code implementing the serviceUrl and itemUrl adapter in the Open VSX project:

https://github.com/eclipse/openvsx/tree/f653161c7bd9b08ae80ba24df758839d72bc0fe0/server/src/main/java/org/eclipse/openvsx/adapter

The corresponding urls are:
```json
"extensionsGallery": { "serviceUrl": "https://open-vsx.org/vscode/gallery", "itemUrl": "https://open-vsx.org/vscode/item" }
```

ShamrockLee commented:
I originally plan to extract extensions from the `index.json` fetch by the obsolete cacheUrl (https://vscode.blob.core.windows.net/gallery/index) so that we can have 2000+ new extensions from the Marketplace.
Now I think we had better drop it and prevent any of the derived work from entering the source tree due to the EULA of the VSCode Marketplace.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/in-tree-or-out-of-tree/18389/1

ShamrockLee commented:
I just encountered something called **extension packs**, which aren't managed properly by the current `vscode-extensions` framework.

It's quite easy to get the information -- by fetching the attribute `extensionPack` from `package.json` when bootstrapping / updating `registry-reference-list.json`, but we need to find a way to inject them into the VSCode / Codium.

schuelermine commented:
hey @ShamrockLee are you still working on this? I recently thought of the same thing so I opened #197293, and I’m wondering if I should close in favor of this PR.

ShamrockLee commented:
This one is still relevant, and it should already work on extensions that doesn't require custom build process.

There are some problems, through:

*   I wonder if the overlay-style build process abstraction in this PR would be over-enginneered. Not sure if it shoud be simplified somehow.

*   The custom build process (for extensions which needs more than unpack-and-installing to get built) needs to be ported manually to the new framework. It could be shared between registries once the name issues mentioned below could be resolved.

*   Extensions are identified with their `publisher.name`, which is generally the same accross registries. The package attribute name of each extension is usually identical to its `publisher.name`. However, the exceptions are not neglectable. One example of the name-nad-attribute mismatch is the package identified as `"bbenoist.Nix"` being renamed from `vscode-extensions.bbenoist.Nix` to `vscode-extensions.bbenoist.nix`. Should we just insist that the attribute be identical to `publisher.name` in this new framework? The naming difference between registries should be a rather mild problem, as custom build process can be mapped to different names.

schuelermine commented:
> Should we just insist that the attribute be identical to `publisher.name` in this new framework?

I think that sounds reasonable. Couldn’t any other things be done via the aliases in the current system anyway?

ShamrockLee commented:
The attribute renaming wasvs made in commit 9ee8bd188933750be0584f285daf9a295d0c8930 . It casted the letter to lower case to "keep the naming convention uniform".

schuelermine commented:
that doesn’t sound reasonable. IMO if the source is inconsistent it’d be inconsistent to make it consistent

ShamrockLee commented:
@schuelermine I didn't even realize that `.vsix` files are now provided per-platform before your PR. Do you know how the platform-specific packages differ from the ones downloaded without specifying the platform? BTW, is there a way to download .vsix files platform-wise from VSCode Marketplace?

schuelermine commented:
No, I don't know anything about it. I just copied the platform distinction from the API reference for Open VSX and I reused the arch property from the marketplace code, assuming it refers to the same thing.

Im-Beast commented:
Are there any updates on this?

ShamrockLee commented:
On 02.08.2023 15:44, Mateusz Franik wrote:
> Are there any updates on this?
> 

Sorry for disappearing for so long.

Here's the things done:

1.  `registry-references-list.json` from OpenVSX can be generated 
programmatically using `nix-prefetch-openvsx`.
2.  `registry-references-list.json` from VSCode Marketplace are 
generated manually, but is probably outdated.
3.  All the extensions from `registry-references-list.json` that doesn't 
need special build process can be built.

Todo:

1.  Unite the naming convention. Quite some extension attributes are 
changed to lower case instead of following the "extension publisher" and 
"extension name" from the upstream. I personally think it's a good idea 
to rename them back.
2.  Port all the custom build process to to the new framework. This 
requires a lot of manual work against a moving target (the current 
extension set), and is the main blocker of this PR.
3.  The old VSCode marketplace asset URLs cannot fetch some new 
extensions, and the relevant URLs seem to come with certain filters that 
blocks direct `curl`s. This affects not only this PR and the 
`nix-prefetch-vscode-marketplace` prefetcher, but also the current 
extension set (built with 
`vscode-utils.buildVscodeMarketplaceExtension`).
     We would need to solve this with VSCode-like User Agent, or even 
custom headers. This require inspecting against the VSCode OSS source 
code.
4.  Automate production of the `registry-references-list.json` from the 
current extension set. This requires the hash-conversion builtin 
function (https://github.com/NixOS/nix/pull/7708). With such builtin 
function, we could also convert SHA256 hash fetched from OpenVSX without 
actually downloading the VSIX file and compute the hash.
5.  Simplify the overlay structure. The one in this PR seems 
over-engineered.

> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> 
> Links:
> ------
> [1] 
> https://github.com/NixOS/nixpkgs/pull/121583#issuecomment-1662241041
> [2]
> https://github.com/notifications/unsubscribe-auth/AKQF2M2HUAT2M66WHRV7W3DXTJKSLANCNFSM44AP6DBA
